### ISSUE TITLE
Images can be duplicated if multiple people upload them

### ISSUE THREAD

**Issue by sarayourfriend** [WordPress/openverse-catalog#490]
## Description

These images will have different URLs, different foreign identifiers, but the content of the image will be the exact same

Originally reported by @AetherUnbound 

## Additional context

Madison suggested we could use a perceptual hash of each image to try to catch duplicates.

How would we decide which one to keep?

## Resolution

- [ ] 🙋 I would be interested in resolving this bug.

**Comment by thedevhaider** [comment]
@sarayourfriend Yeah, this makes sense. There can be a situation of duplicate images if multiple people upload them. I came across one open-source library [imagehash](https://github.com/JohannesBuchner/imagehash) that implements a perceptual hashing algorithm to compare images by calculating hashes of different images. 
There are basically average hashing and perceptual hashing algorithms used for different purposes. Refer to this [link](https://www.hackerfactor.com/blog/index.php?/archives/432-Looks-Like-It.html) to understand the example implementation. As per the above link.

> Average Hash is a great algorithm if you are looking for something specific. For example, if I have a small thumbnail of an image and I know that the big one exists somewhere in my collection, then Average Hash will find it very quickly. However, if there are modifications -- like text was added or a head was spliced into place, then Average Hash probably won't do the job. While pHash is slower, it is very tolerant of minor modifications (minor being less than 25% of the picture).

Refer to this [link](https://ourcodeworld.com/articles/read/1006/how-to-determine-whether-2-images-are-equal-or-not-with-the-perceptual-hash-in-python) to compare two images using the perceptual algorithm. 

As there will be duplicate images so keeping which image becomes a challenge. We could do something like keeping an image that is better than the other same images. 

An image can be better based on the below parameters.

1. As per the [article](https://www.hackerfactor.com/blog/index.php?/archives/432-Looks-Like-It.html), with pictures, high frequencies give you detail, while low frequencies show you structure. A large, detailed picture has lots of high frequencies. A very small picture lacks details, so it is all low frequencies. So the image that has a higher frequency will be the winner. 
2. The image has a higher resolution.
3. The image has better contrast and colors.

**Comment by sarayourfriend** [comment]
Thanks for sharing your thoughts on this @thedevhaider! That sounds like a pretty reasonable approach to _automating_ choosing between two uploaders. Maybe that outcome would also tend to pick maybe the "original" artist over someone re-uploading. Another interesting thing we could do in this case is flag the image as clashing with an existing perceptual hash and then try to communicate with the uploaders of the different images. Maybe sometimes it will be the same person on different platforms, but we could also potentially catch some cases of license violations :thinking: That could get us into some weird "enforcement" territory though and I don't think the project necessarily wants to be there. It might be safest for us to just soft-delete duplicates from the catalog and not surface them in the API... and then decide sometime down the road what we'd want to do about it.

One curious case of this, however, which would be complicated, is how to handle with people uploading pictures of museum artwork. Will nearly identical pictures of the same Van Gogh get flagged? Does this mean we'd have zero relevant results for ["Van Gogh starry night"](https://wordpress.org/openverse/search/?q=van%20gogh%20starry%20night)?

Even if we left the detected "highest quality" one, is it fair for us to automatically decide that? And would we have to mix the license into the perceptual hash for these cases? How would we detect these case instead of an image that had been uploaded twice under two different licenses?

This ticket has a _ton_ of interesting and difficult questions for us to work through!

**Comment by thedevhaider** [comment]
@sarayourfriend Yes, I agree with your first point. Meanwhile, there will be a discussion going on for different paths we can take we could soft delete the duplicate images based on some logic (Maybe perceptual or something else). 

You have raised a valid point regarding the museum artwork or any artwork where there can be similarities. This will need some discussions as there could be different ways of resolving this. one of the ways I could think of is to treat uploaded artwork images as NFTs but I am not sure if this territory we can touch or how to take it ahead. 

I have not much to say about the third point, but this seems similar to your second point. Maybe NFTs in this case as well help us to figure out this problem. Due to my limited knowledge of NFT contracts, I don't have much to say.

**Comment by sarayourfriend** [comment]
I don't think NFTs can help us in this case. NFTs do not hold the actual work they represent, but rather a pointer to it (usually a URL of some sort). Even if they did, we can hash works without needing to implement "smart contracts" or any of the other blockchain/crypto things that are far outside the scope of this project.

I think we also want to stray away from anything that would get us into the territory of license enforcement. It's not one of the goals of the Openverse project, now or in the future, as far as I know. I think potentially the best thing we could do is something like other image search providers and group images that have similar perceptual hashes into groups, so that when you click on the group, it opens up a secondary results view of all the results that fall within that perceptual hash. In this mode, anyone could upload their own pictures of the same famous painting and they would all be represented in Openverse under the same grouped result.

I don't know what the technical details of that would be but it sounds like a big undertaking to me, including probably developing an entirely different result endpoint that supports this as I don't believe we could responsibly change the behavior of the existing endpoints in this way. That is to say, we'd have to solve API versioning or introduce an entirely separate endpoint for this as well as a new index into Elasticsearch. We'd also have to decide how and when to hash the images. We do not currently download the images during ingestion in the catalog, but hashing would require that. We have other metadata that we want to pull from works as well (format, specific sizes, etc).

My point here is that this project intersects with a lot of other things we have going on. I don't think we can prioritize this particular issue now, but there are other things we can do to make this one possible. Probably the most high value is to decide how we can download works during ingestion so that we can extract relevant metadata.

There is also [this semi-related issue](https://github.com/WordPress/openverse-api/issues/654) that intersects with this issue in a variety of ways. If you want to help move the discussion there forward that would be helpful. I think whatever ends up happening in that related issue will enable and inform how we approach this one.
